### PR TITLE
create empty resultIterator

### PR DESCRIPTION
**Pull request by Kharhamel**

### CODE REVIEW

File: src/InnerResultIterator.php [hunk 1]
```diff
@@ -65,23 +66,29 @@ class InnerResultIterator implements \Iterator, \Countable, \ArrayAccess
...
     private $logger;
 
+    protected $count = null;
+
     /**
      * @param mixed[] $parameters
      * @param array[] $columnDescriptors
      */
-    public function __construct(string $magicSql, array $parameters, ?int $limit, ?int $offset, array $columnDescriptors, ObjectStorageInterface $objectStorage, ?string $className, TDBMService $tdbmService, MagicQuery $magicQuery, LoggerInterface $logger)
+    public function __construct(?string $magicSql, ?array $parameters, ?int $limit, ?int $offset, ?array $columnDescriptors, ?ObjectStorageInterface $objectStorage, ?string $className, ?TDBMService $tdbmService, ?MagicQuery $magicQuery, ?LoggerInterface $logger)
```
Comment (moufmouf) [inline src/InnerResultIterator.php]: Idea: we really need 2 different constructors here. We could put the constructor private and empty, and create 2 static methods to create an instance.

File: src/PageIterator.php [hunk 2]
```diff
@@ -71,7 +72,7 @@ class PageIterator implements Page, \ArrayAccess, \JsonSerializable
      * @param mixed[] $parameters
      * @param array[] $columnDescriptors
      */
-    public function __construct(ResultIterator $parentResult, string $magicSql, array $parameters, int $limit, int $offset, array $columnDescriptors, ObjectStorageInterface $objectStorage, ?string $className, TDBMService $tdbmService, MagicQuery $magicQuery, int $mode, LoggerInterface $logger)
+    public function __construct(ResultIterator $parentResult, ?string $magicSql, array $parameters, ?int $limit, ?int $offset, ?array $columnDescriptors, ?ObjectStorageInterface $objectStorage, ?string $className, ?TDBMService $tdbmService, ?MagicQuery $magicQuery, ?int $mode, ?LoggerInterface $logger)
```
Comment (moufmouf) [inline src/PageIterator.php]: Same comment as above. We could put the constructor private and empty, and create 2 static methods to create an instance.

File: src/ResultIterator.php [hunk 3]
```diff
@@ -69,20 +70,29 @@ class ResultIterator implements Result, \ArrayAccess, \JsonSerializable
     /**
      * @param mixed[] $parameters
      */
-    public function __construct(QueryFactory $queryFactory, array $parameters, ObjectStorageInterface $objectStorage, ?string $className, TDBMService $tdbmService, MagicQuery $magicQuery, int $mode, LoggerInterface $logger)
+    public function __construct(?QueryFactory $queryFactory, ?array $parameters, ?ObjectStorageInterface $objectStorage, ?string $className, ?TDBMService $tdbmService, ?MagicQuery $magicQuery, int $mode, ?LoggerInterface $logger)
```
Comment (moufmouf) [inline src/ResultIterator.php]: Also same comment.

File: src/MapIterator.php [hunk 4]
```diff
@@ -55,7 +55,7 @@ public function __construct($iterator, callable $callable)
      * Alters the current item with $this->callable and returns a new item.
      * Be careful with your types as we can't do static type checking here!
      *
-     * @return mixed
+     * @return mixed|null
```
Comment (moufmouf) [inline src/MapIterator.php]: Mixed can be null to? Can we revert to:

```suggestion
     * @return mixed
```

File: src/InnerResultIterator.php (hunk 1 above)
Comment (Kharhamel) [reply src/InnerResultIterator.php]: done

File: src/PageIterator.php (hunk 2 above)
Comment (Kharhamel) [reply src/PageIterator.php]: done

File: src/ResultIterator.php (hunk 3 above)
Comment (Kharhamel) [reply src/ResultIterator.php]: done

**Pull request merged**